lib/src/config_document_parser.cc

```diff
@@ -0,0 +1,632 @@
+            if (t->get_token_type() == token_type::IGNORED_WHITESPACE || t->get_token_type() == token_type::NEWLINE
+                    || is_unquoted_whitespace(t)) {
+                nodes.push_back(make_shared<config_node_single_token>(t));
+                if (t->get_token_type() == token_type::NEWLINE) {
+                    _line_number = t->line_number() + 1;
+                }
+            } else if (t->get_token_type() == token_type::COMMENT) {
+                nodes.push_back(make_shared<config_node_comment>(t));
+            } else {
+                if (t->line_number() >= 0) {
+                    _line_number = t->line_number();
+                }
+                return t;
+            }
+        }
+    }
+
+    void parse_context::put_back(shared_token token) {
+        _buffer.push(token);
+    }
+
+    bool parse_context::check_element_separator(shared_node_list& nodes) {
+        if (_flavor == config_syntax::JSON) {
+            shared_token t = next_token_collecting_whitespace(nodes);
+            if (t->get_token_type() == token_type::COMMA) {
+                nodes.push_back(make_shared<config_node_single_token>(t));
+                return true;
+            } else {
+                put_back(t);
+                return false;
+            }
+        } else {
+            bool saw_separator_or_newline = false;
+            shared_token t = next_token();
+            while (true) {
+                if (t->get_token_type() == token_type::IGNORED_WHITESPACE || is_unquoted_whitespace(t)) {
+                    nodes.push_back(make_shared<config_node_single_token>(t));
+                } else if (t->get_token_type() == token_type::COMMENT) {
+                    nodes.push_back(make_shared<config_node_comment>(t));
+                } else if (t->get_token_type() == token_type::NEWLINE) {
+                    saw_separator_or_newline = true;
+                    _line_number++;
+                    nodes.push_back(make_shared<config_node_single_token>(t));
+                    // we want to continue to also eat a comma if there is one
+                } else if (t->get_token_type() == token_type::COMMA) {
+                    nodes.push_back(make_shared<config_node_single_token>(t));
+                    return true;
+                } else {
+                    // non-newline-or-comma
+                    put_back(t);
+                    return saw_separator_or_newline;
+                }
+                t = next_token();
+            }
+        }
+    }
+
+    shared_node_value parse_context::consolidate_values(shared_node_list &nodes) {
+        // this trick is not done in JSON
+        if (_flavor == config_syntax::JSON) {
+            return nullptr;
+        }
+
+        // create only if we have value tokens
+        shared_node_list values;
+        int value_count = 0;
+
+        // ignore a newline up front
+        shared_token t = next_token_collecting_whitespace(nodes);
+        while (t) {
```

Comment:
`while(t)` instead of `while(true)`? Did this come up as a bug, because it differs from https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/ConfigDocumentParser.java#L141?